python/mlx/nn/layers/upsample.py

```diff
@@ -0,0 +1,145 @@
+# Copyright © 2023-2024 Apple Inc.
+
+from typing import List, Literal, Tuple, Union
+
+import mlx.core as mx
+from mlx.nn.layers.base import Module
+
+
+def upsample2d_nearest(x: mx.array, scale: Tuple[float, float]):
+    # Integer scales means we can simply expand-broadcast and reshape
+    if tuple(map(int, scale)) == scale:
+        sh, sw = map(int, scale)
+        B, H, W, C = x.shape
+        x = x[:, :, None, :, None]
+        x = mx.broadcast_to(x, (B, H, sh, W, sw, C))
+        x = x.reshape(B, H * sh, W * sw, C)
+        return x
+
+    # Floating point scale means we need to do indexing
+    else:
+        sh, sw = scale
+        B, H, W, C = x.shape
+        new_H = int(H * sh)
+        new_W = int(W * sw)
+        idx_y = (mx.arange(0, new_H) / sh).astype(mx.int32)
+        idx_x = (mx.arange(0, new_W) / sw).astype(mx.int32)
+        return x[:, idx_y[:, None], idx_x[None]]
+
+
+def upsample2d_bilinear(x: mx.array, scale: Tuple[float, float]):
+    sh, sw = scale
+    B, H, W, C = x.shape
+    new_H = int(H * sh)
+    new_W = int(W * sw)
+    idx_y = mx.arange(0, new_H) * ((H - 1) / (new_H - 1))
+    idx_x = mx.arange(0, new_W) * ((W - 1) / (new_W - 1))
+    # Compute the sampling grid
+    idx_y_t = mx.floor(idx_y).astype(mx.int32)
+    idx_y_b = mx.ceil(idx_y).astype(mx.int32)
+    idx_x_l = mx.floor(idx_x).astype(mx.int32)
+    idx_x_r = mx.ceil(idx_x).astype(mx.int32)
+    # Sample
+    a = x[:, idx_y_t[:, None], idx_x_l[None]]
+    b = x[:, idx_y_t[:, None], idx_x_r[None]]
+    c = x[:, idx_y_b[:, None], idx_x_l[None]]
+    d = x[:, idx_y_b[:, None], idx_x_r[None]]
+    # Compute bilinear interpolation weights
+    y_weight = (idx_y - idx_y_t)[:, None, None]
+    x_weight = (idx_x - idx_x_l)[None, :, None]
+    w_a = (1 - x_weight) * (1 - y_weight)
+    w_b = x_weight * (1 - y_weight)
+    w_c = y_weight * (1 - x_weight)
+    w_d = x_weight * y_weight
+    # Interpolate
+    return w_a * a + w_b * b + w_c * c + w_d * d
+
+
+class Upsample2d(Module):
```

Comment:
I notice PyTorch has a single `Upsample` class which handles different dimensions. It might be worth making that consistent and then throwing (or supporting) on the dimensions not yet handled.

Reply:
https://pytorch.org/docs/stable/generated/torch.nn.Upsample.html

Reply:
@awni Thank you for this suggestion. I was thinking of implementing upsampling within `nn.Upsample`. I had a similar idea/comment on https://github.com/ml-explore/mlx/pull/357.  There, I went with `nn.Pooling` instead of `nn.MaxPooling1d` or `nn.MaxPooling2d` and @angeloskath suggested we go for a different class based on the dimension or pooling type.  Thus, to be consistent with that review, I implemented ``nn.Upsample2d``.  In my opinion, 2D upsampling is also the most common use case. 
Could you please share your thoughts on whether we want `nn.Upsample2d` or `nn.Upsample`, based on what we might have for pooling?